File: 02_insertion.c
#include <stdio.h>
#include <stdlib.h>

struct Node
{
	int index;
	char data;
	struct Node *next;
	struct Node *prev;
};

int Llength(struct Node* list)
{
	int count = 0;
	while (list != NULL){
		list = list->next;
		count++;
	}
	return count;
}

struct Node* findNode(struct Node** head_r, char* q_data)
{
	struct Node *cursor = *head_r;
	while (cursor != NULL){
		if (cursor->data == *q_data){
			return cursor;
		} else {
			cursor = cursor->next;
		}
	}
	printf("Requested Value not found in List using findNode.");
}

struct Node* findHead(struct Node* current)
{
	while(current->prev != NULL){
		current = current->prev;
	}
	return current;
}

struct Node* findTail(struct Node* current)
{
	while(current->next != NULL){
		current = current->next;
	}
	return current;
}

void incrementFollowing(struct Node* increase)
{
	// set index to the node we want to increase from
	int increment = increase->index;
	struct Node *temp = increase->next;
	while (temp != NULL){
		increment = increment + 1;
		temp->index = increment;
		temp = temp->next;		
	}
}

void decrementFollowing(struct Node* decrease)
{
	// set index to the node we want to increase from
	int decrement = decrease->index;
	struct Node *temp = decrease->next;
	while (temp != NULL){
		decrement = decrement - 1;
		temp->index = decrement;
		temp = temp->next;		
	}
}

void push(struct Node** head_r, char* new_data)
{
	struct Node* new_node = (struct Node*) malloc(sizeof(struct Node));
	
	new_node->prev = NULL;
	new_node->index = 0;		// set index to 0
	new_node->data = *new_data;	// push new data
	new_node->next = (*head_r);	// set next to head
	(*head_r)->prev = new_node;	// set prev of head to new
	(*head_r) = new_node;		// set current head to new
	incrementFollowing(new_node);   // increment index of following 
}

void append(struct Node** head_r, char* new_data)
{
	struct Node* new_node = (struct Node*) malloc(sizeof(struct Node));
	struct Node *last = *head_r;

	new_node->data = *new_data;
	new_node->next = NULL;

	if (*head_r == NULL)
	{
		*head_r = new_node;
		return;
	}

	while (last->next != NULL)
		last = last->next;

	new_node->prev = last;
	new_node->index = last->index + 1;
	last->next = new_node;
	return;
}

void insertAfter(struct Node* prev_n, char* new_data)
{
	if (prev_n == NULL)
	{
		printf("the given previous node cannot be NULL");
		return;
	}

	struct Node* new_node = (struct Node*) malloc(sizeof(struct Node));
	new_node->data  = *new_data;	// set data of new
	new_node->next = prev_n->next;	// set next to the prev next
	new_node->prev = prev_n;
	prev_n->next = new_node;	// set next of prev to current
	struct Node *next_n = new_node->next;		
	next_n->prev = new_node;	// set prev of next to current
	new_node->index = next_n->index;// set index of new node to next index
	incrementFollowing(new_node); 	// increment index after current node
}

void insertBefore(struct Node* next_n, char* new_data)
{
	if (next_n == NULL){
		printf("The given next node cannot be null.");
		return;
	}

	if (next_n->prev == NULL) {
		push(&next_n, new_data);
		return;
	}
	
	struct Node* new_node = (struct Node*) malloc(sizeof(struct Node));
	new_node->data = *new_data;
	new_node->prev = next_n->prev;
	new_node->next = next_n;
	next_n->prev = new_node;
	struct Node* prev_n = new_node->prev;
	prev_n->next = new_node;
	new_node->index = next_n->index;
	incrementFollowing(new_node);
}

void printL(struct Node *node)
{
	while (node != NULL){
		printf("%d - %c\n", node->index, node->data);
		node = node->next;
	}
}

int main()
{
	struct Node* head = NULL;
	append(&head, "D");
	push(&head, "B");
	push(&head, "A");
	append(&head, "E");

	insertAfter(findNode(&head, "B"), "C");

	append(&head, "G");
	insertBefore(findNode(&head, "G"), "F");

	printf("Linked List contains: \n");
	printL(head);

	printf("\nLeng: %d \n", Llength(head));

	return 0;
}

File: linked_lists/02_insertion.py
class Node:
    def __init__(self, data):
        self.index = 0
        self.data = data
        self.next = None
        self.prev = None

class LList:
    def __init__(self):
        self.head = None
    
    def find(self, item):
        if self.head is None:
            print ("Nothing in list.")
            return
        
        search = self.head
        while (search):
            if search.data == item:
                return search
            search = search.next

        print(item + " not found in list.")
        return

    def length(self):
        if self.head is None:
            return 0

        last = self.head
        while (last.next):
            last = last.next

        size = last.index + 1
        return size

    def incrementAfter(self, new_node):
        increase = new_node
        
        while (increase.next):
            increase = increase.next
            increase.index = increase.index + 1

    def decrement(self, new_node):
        increase = new_node
        
        while (increase.next):
            increase = increase.next
            increase.index = increase.index - 1

    def push(self, new_data):
        new_node = Node(new_data)

        # new.next is the head
        new_node.next = self.head

        self.head.prev = new_node
        
        self.incrementAfter(new_node)

        self.head = new_node

    def append(self, new_data):
        new_node = Node(new_data)
        
        # if list is empty, new node is head
        if self.head is None:
            self.head = new_node
            return

        # else move to the end of the list
        last = self.head
        while (last.next):
            last = last.next
        
        # assign the new.prev to last
        # and last.next to new
        new_node.prev = last
        new_node.index = last.index + 1
        last.next = new_node
 
    def insertBefore(self, next_node, new_data):
        next_node = self.find(next_node)

        if next_node is None:
            print("Invalid previous node.")
            return

        if next_node.prev is None:
            self.push(new_data)
        else:
            new_node = Node(new_data)
            new_node.prev = next_node.prev
            new_node.index = next_node.index
            new_node.next = next_node
            next_node.prev = new_node
            prev = new_node.prev
            prev.next = new_node
            self.incrementAfter(new_node)


    def insertAfter(self, prev_node, new_data):
        prev_node = self.find(prev_node)

        if prev_node is None:
            print("Invalid previous node.")
            return

        new_node = Node(new_data)
        new_node.next = prev_node.next
        new_node.index = prev_node.index + 1

        self.incrementAfter(new_node)

        prev_node.next = new_node

    def out(self):
        temp = self.head
        while (temp):
            print(str(temp.index) + " " + temp.data)
            temp = temp.next

if __name__=='__main__':
    linked = LList()

    linked.append("D")
    linked.push("B")
    linked.push("A")
    linked.append("E")
    linked.insertAfter("B", "C")
    linked.append("G")
    linked.insertBefore("G", "F")
    linked.out()
    print(linked.length())
